// @adonisjs/sink/src/Files/index.js
"use strict";
/*
 * @adonisjs/files
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageJsonFile = exports.EnvFile = exports.AdonisRcFile = exports.YamlFile = exports.TemplateLiteralFile = exports.NewLineFile = exports.MustacheFile = exports.JsonFile = exports.IniFile = exports.KeyValuePair = exports.File = void 0;
var File_1 = require("./Base/File");
Object.defineProperty(exports, "File", { enumerable: true, get: function () { return File_1.File; } });
var KeyValuePair_1 = require("./Base/KeyValuePair");
Object.defineProperty(exports, "KeyValuePair", { enumerable: true, get: function () { return KeyValuePair_1.KeyValuePair; } });
var Ini_1 = require("./Formats/Ini");
Object.defineProperty(exports, "IniFile", { enumerable: true, get: function () { return Ini_1.IniFile; } });
var Json_1 = require("./Formats/Json");
Object.defineProperty(exports, "JsonFile", { enumerable: true, get: function () { return Json_1.JsonFile; } });
var Mustache_1 = require("./Formats/Mustache");
Object.defineProperty(exports, "MustacheFile", { enumerable: true, get: function () { return Mustache_1.MustacheFile; } });
var NewLine_1 = require("./Formats/NewLine");
Object.defineProperty(exports, "NewLineFile", { enumerable: true, get: function () { return NewLine_1.NewLineFile; } });
var TemplateLiteral_1 = require("./Formats/TemplateLiteral");
Object.defineProperty(exports, "TemplateLiteralFile", { enumerable: true, get: function () { return TemplateLiteral_1.TemplateLiteralFile; } });
var Yaml_1 = require("./Formats/Yaml");
Object.defineProperty(exports, "YamlFile", { enumerable: true, get: function () { return Yaml_1.YamlFile; } });
var AdonisRc_1 = require("./Special/AdonisRc");
Object.defineProperty(exports, "AdonisRcFile", { enumerable: true, get: function () { return AdonisRc_1.AdonisRcFile; } });
var Env_1 = require("./Special/Env");
Object.defineProperty(exports, "EnvFile", { enumerable: true, get: function () { return Env_1.EnvFile; } });
var PackageJson_1 = require("./Special/PackageJson");
Object.defineProperty(exports, "PackageJsonFile", { enumerable: true, get: function () { return PackageJson_1.PackageJsonFile; } });
